### ISSUE TITLE
Hook not executed for netctl-auto

### ISSUE THREAD

**Issue by vitoreiji** [joukewitteveen/netctl#185]
I created a hook that is executed just fine when I do `netctl start my_profile` but it is not executed at all when I `systemctl start netctl-auto@<interface>`. I thought it was supposed to be executed for each profile that is eligible for auto starting.

## To reproduce:

    # cat /etc/netctl/my_profile
    Interface=wlp2s0
    Connection=wireless
    Security=wpa
    ESSID=MyWifi
    IP=dhcp
    Key="hunter2"

    # cat /etc/netctl/hooks/my_hook
    echo Hook ran for profile $Profile >> /root/hook_trace

    # ls -l /etc/netctl/hooks/my_hook
    total 4
    -rwxr-xr-x 1 root root 55 may  1 16:17 /etc/netctl/hooks/my_hook

    # systemctl start netctl-auto@wlp2s0.service

    # cat /root/hook_trace
    cat: /root/hook_trace: No such file or directory

    # netctl start my_profile

    # cat /root/hook_trace
    Hook ran for profile my_profile

**Comment by joukewitteveen** [comment]
Your observation is correct: `netctl-auto` does not source hooks. This is by design, since hooks are supposed to support the transition from down to routable of an interface (i.e. what `netctl start` does). With `netctl-auto`, however, we can transition from one profile to another, and won't be notified of it. This leaves a few possibilities for this bug.

1. Find a meaningful way to use hooks in `netctl-auto`. I don't think this has a high chance of success.
2. Update the documentation to more clearly state that hooks are not effective for `netctl-auto`.
3. Do nothing to `netctl`, consider your use case, and try to find a better way to support it.

In the third case, if, for instance, you are trying to call some program when an interface is configured, you could also write a systemd service for your program and order it after `network.target`. In this particular case, that would be a more generic solution. For your particular use case, another solution may or may not be available.

**Comment by vitoreiji** [comment]
Thanks for the explanation! My use case is setting the `Key` variable from an external program (`pass` in my case). My plan was to call `pass "$Profile"` from the hook, so that I can separate the keys from the profiles, so that I can put my profiles in a git repo and not worry about it.

If you don't think this is a use case you're interested in supporting, then I'd say update the documentation (and the [Arch Wiki](https://wiki.archlinux.org/title/Netctl#Using_hooks)), and I'll keep doing what I'm doing right now, which is setting `Key="$(pass ${Interface}-${ESSID})"` in each profile file. I wanted to move this to a hook because the manual says these files "are not expected to execute any code", so I thought this might just stop working at some point.

**Comment by joukewitteveen** [comment]
That is an excellent use case if you ask me. Looking at the code, the hooks are sourced only when the connection status changes, which is too late to do what you want to do. However, loading the hooks when first scanning the profile files on `netctl-auto start` is not possible to do in a way that meets the documented behavior. From `netctl.profile(5)`:
>  The name of the current profile is available via the Profile variable in the environment of the sourced scripts. On starting or stopping a profile, the respective command is available via the Command variable.

This fragment is already not true for interface scripts, which are more liberally sourced. I am inclined to change "scripts" into "hooks". I would also prefer not to source the hooks on `netctl-auto start`. Even though sourcing the hooks at that point would be easy to implement, I think doing so would blur the logic even further than it is already, and potentially introduces nasty side effects. Where in the man pages would you be looking if you wanted to know whether hooks are sourced on `netctl-auto start`?

Returning to your use case, any function you were thinking of you could add to an interface script. Granted, that would require duplicating the code per interface, but it might be preferable to duplicating per profile. On the other hand, I expect `Key="$(pass ${Interface}-${ESSID})"` to keep working forever, so if you have a suggestion for toning down that part of the documentation, I would love to hear it too.

**Comment by vitoreiji** [comment]
> potentially introduces nasty side effects

I can definitely see that.

> Where in the man pages would you be looking if you wanted to know whether hooks are sourced on `netctl-auto start`?

Either where hooks are currently described, [on the second paragraph of the Description in `netctl.profile.5`](https://github.com/joukewitteveen/netctl/blob/9b70046f86b948c7e8c3c307f218c378714692ed/docs/netctl.profile.5.txt#L24-L27) or anywhere on `netctl-auto.1`. What threw me off was that the current description says:

> *Whenever a profile is read*, all executable scripts in **/etc/netctl/hooks/** and any executable script in **/etc/netctl/interfaces/** with the name of the interface for the profile are sourced.

In my mind, starting `netctl-auto` would read all profiles at least to figure out which interface they're using. And I guess it does, just in this case the scripts aren't called, so I would change "whenever a profile is read" to "when the connection status changes".

> Returning to your use case, any function you were thinking of you could add to an interface script.

I tried that too, but no cigar. The script is called twice (I have a couple dozen profiles) and none of the documented variables, like `$Profile`,  seem to be set:

```bash
ACTION_SCRIPT=/usr/lib/netctl/auto.action
BASH=/bin/bash
BASHOPTS=checkwinsize:cmdhist:complete_fullquote:extquote:force_fignore:globasciiranges:hostcomplete:interactive_comments:progcomp:promptvars:sourcepath
BASH_ALIASES=()
BASH_ARGC=([0]="1" [1]="2")
BASH_ARGV=([0]="/etc/netctl/interfaces/wlp2s0" [1]="wlp2s0" [2]="stop")
BASH_CMDS=()
BASH_LINENO=([0]="9" [1]="286" [2]="0")
BASH_SOURCE=([0]="/etc/netctl/interfaces/wlp2s0" [1]="/usr/lib/netctl/interface" [2]="/usr/bin/netctl-auto")
BASH_VERSINFO=([0]="5" [1]="1" [2]="4" [3]="1" [4]="release" [5]="x86_64-pc-linux-gnu")
BASH_VERSION='5.1.4(1)-release'
DIRSTACK=()
EUID=0
FUNCNAME=([0]="source" [1]="load_interface_config" [2]="main")
GROUPS=()
HOSTNAME=apolo
HOSTTYPE=x86_64
IFS=$' \t\n'
INVOCATION_ID=22cb377998fd4a8a929f0732fb715149
JOURNAL_STREAM=8:312130
LANG=pt_BR.UTF-8
MACHTYPE=x86_64-pc-linux-gnu
NETCTL_VERSION=1.24
OPTERR=1
OPTIND=1
OSTYPE=linux-gnu
PATH=/usr/local/sbin:/usr/local/bin:/usr/bin
PIPESTATUS=([0]="0")
PPID=1
PROFILE_DIR=/etc/netctl
PS4='+ '
PWD=/
SERVICE_RESULT=success
SHELL=/bin/bash
SHELLOPTS=braceexpand:hashall:interactive-comments
SHLVL=1
STATE_DIR=/run/netctl
STATE_FILE=/var/lib/netctl/netctl.state
SUBR_DIR=/usr/lib/netctl
SYSTEMD_EXEC_PID=64621
TERM=dumb
UID=0
_=config_file=/etc/netctl/interfaces/wlp2s0
config_file=/etc/netctl/interfaces/wlp2s0
# Plus the whole script is printed below, not sure why...
```

> if you have a suggestion for toning down that part of the documentation, I would love to hear it too.

How about this:

> The files consist of variable definitions following the bash shell syntax, but executing code from them is discouraged, since they will be sourced at unexpected times.

In any case, I was hoping the hook script would work, because then I could package this together with some helper scripts to migrate all your keys to out of the profiles and distribute it, but *even better* would be the profiles supporting this functionality natively :wink: .

Thanks for considering!

**Comment by joukewitteveen** [comment]
Just to give you an update: I think I can make it work, but I will have to draft and benchmark something. Hopefully, I'll get to that this weekend.

**Comment by joukewitteveen** [comment]
Would you be able to build and patch `netctl` yourself? I would love to know whether the patch on https://lists.archlinux.org/pipermail/arch-projects/2021-May/005429.html works for you. If building manually is too inconvenient, I can send you a packaged version.

**Comment by vitoreiji** [comment]
It works! \o/

Thank you very much for this!

**Comment by joukewitteveen** [comment]
I modified the patch slightly (8773d7f02d394a27a173e74abf5b35219055de4e) and released netctl 1.25. Today, the new release hit `[Testing]` in Arch Linux, so it will probably enter `[Core]` in one or two weeks.

@vitoreiji: Thank you for your contribution to `netctl`!
